### ISSUE TITLE
Offer only active or valid vouchers - bug/improvement

### ISSUE THREAD

**Issue by PeNov** [PrestaShop/PrestaShop#22442]
#### Describe the bug

In the case of many inactive coupons or coupons with the same name for example free shipping is very difficult chose correct discount in the list of deactivated vouchers.

#### Expected behavior

Show only active or valid vouchers. Eventually show label (inactive, outdated, voucher code...).  

#### Steps to Reproduce

1. Go to 'Catalog, Cart Rules, Discounts'
2. Create more discounts, some deactivate
3. Try set compatibility with other discounts
4. See error: Inactive/outdated vouchers are showed...

**Screenshots**

![image](https://user-images.githubusercontent.com/38213520/102330700-2c86ca80-3f8a-11eb-8041-81e6a654b145.png)

![image](https://user-images.githubusercontent.com/38213520/102330722-33154200-3f8a-11eb-8362-250e48f3c3cc.png)

#### Additional information

* PrestaShop version: 1.7.6.7 and 1.7.7.0
* PHP version: 7.0

**Comment by hibatallahAouadni** [comment]
Hello @PeNov 

Are you sugguesting to disable the voucher when it is expired automatically?
If so, it's not a bug but it's an improvement!
Please confirm and feedback.

Thanks!

**Comment by PeNov** [comment]
Hello Hibatallah,
No.

Main problem is that prestashop showing all discounts (expired, un-activate, …) in the list of **Compatibility with other cart rules**.

Second problem is in case of same discount name. Here is missing any description, ID, state.

**Steps:**
1)	Create discount with name FREE SHIPPING for order total more than 50EUR.
2)	Add additional discount in BO for realized order number 1 and 2. Name of the discount: FREE SHIPPING. After adding these discounts to the order 1 and 2, prestashop create two discounts with name FREE SHIPPING and inactivate + expired these discounts.
3)	Now we would like create discount 10% which is possible combinate with Free shipping from point 1)
4)	But Compatibility with other cart rules show all Free shipping discounts. **Which discount is from point 1?** We have not ID, description, state in this list. We must rename discount for easy selection.

Discounts overview:
![image](https://user-images.githubusercontent.com/38213520/102509418-d0ee3700-4086-11eb-93c6-08a4ea3c0137.png)

Here you can see 3 discount with name Free shipping. But I don't know which is from point 1)
![image](https://user-images.githubusercontent.com/38213520/102509441-d8addb80-4086-11eb-9293-f41182403bad.png)

Thank you
Best regards
PN

**Comment by PululuK** [comment]
Duplicate of https://github.com/PrestaShop/PrestaShop/issues/19393

**Comment by hibatallahAouadni** [comment]
Hello @PeNov 

Thanks for the clarification. I managed to reproduce it with PS1.7.7.0. But, I believe like I mentionned above, it's not a bug but it's a PrestaShop limitation.

Ping @PrestaShop/product-team @PeNov suggests we need to add the cart rules id at this stage (see screenshot below):

![image](https://user-images.githubusercontent.com/23308427/102781899-b4a90d80-4398-11eb-98e2-5c1a79c94f7c.png)

Cause, if we have a multiple cart rules with the same name, it will be confused to select the desired one.

PS: @PululuK thanks for you help, #19393 is related to this one but not the same :wink: 

Thanks!

**Comment by prestascott** [comment]
Hello everyone,

Thanks @PeNov for your report. :)

This is an interesting feature but why merchant would use exactly the same cart rules name when he can simply be more specific on the name of the cart rule and use a different one? 🤔 

I don't think just adding the ID of the discount will help in this case, because you will still have to check on the discounts list page if you selected the right discount. So we rather add the description in this case.

**Comment by MatShir** [comment]
The description looks like the best solution. @prestascott Could you update the design for the cart rule page? 😄 I will try to push the feature request with the migration of the page

**Comment by MatShir** [comment]
We could show the state of the discount. For me, the incompatible cart rules should show all the cart rules because if you activate an older cart rule, how the new cart rules will interact with it?

**Comment by prestascott** [comment]
@MatShir I just updated the design for the Cart Rules page and added the Description column right next to the Name column. 
You can find the mockups [here](https://www.figma.com/proto/vFA252C7vHmMKc3gcnMkKw/Scott---Cart-Rules-(migration)?node-id=0%3A2&scaling=min-zoom).

**Comment by hibatallahAouadni** [comment]
@MatShir @prestascott Thanks guys for your hard working 🚀

So I changed the issue status to **`To Do`** if you think that it still need specs just change it to **`TBS`** 😉

Thanks!

**Comment by boo-code** [comment]
Still reproducible on 9.2. `AdminCartRulesController::ajaxProcessLoadCartRules()` builds every entry of both lists as the name and nothing else:

```php
'<option value="' . (int) $cart_rule['id_cart_rule'] . '">&nbsp;' . Tools::safeOutput($cart_rule['name']) . '</option>'
```

Measured with five cart rules, two of them deliberately sharing a name:

```
ZZ Probe disabled
ZZ Probe expired
ZZ Probe scheduled
ZZ Probe live
```

Nothing in that distinguishes the disabled one, the expired one, or the two that share a name - which is the report exactly.

**The conclusion this thread reached is the right one and it is not the title.** MatShir's point stands: compatibility is a lasting relation, so filtering the list would silently drop a relation the merchant then cannot restore - "if you activate an older cart rule, how will the new cart rules interact with it?". So the answer is to label, not to hide.

Two things make that cheap. `CartRule::getAssociatedRestrictions()` already runs `SELECT t.*, tl.*`, so `active`, `code`, `date_from` and `date_to` are on every row the loop walks - the state needs no extra query. And `Disabled`, `Expired` and `Scheduled` already exist in `Admin.Global` and `Admin.Catalog.Feature`, so it needs no new strings either.

One thing worth knowing before anyone spends review time here: **the new discount system does not carry this feature forward.** `DiscountFormDataHandler` sets `setCompatibleDiscountTypeIds(...)` - compatibility by discount *type*, not the per rule list this issue is about. So when `FEATURE_FLAG_DISCOUNT` becomes the default these lists stop existing. That flag is currently `state="0"`, `stability="beta"`, so every shop is still on the legacy page, but it does mean this issue resolves by removal rather than by fix once the migration lands - and it is worth deciding which of those you want before the design work on the new page is finished, since per rule compatibility is being dropped rather than replaced.
